File: src/lib/pets.ts
import clawdAtlasUrl from "../assets/pets/clawd/spritesheet.webp";

export type PetSource = "bundled" | "external";

export interface PetCandidate {
  id: string;
  name: string;
  source: PetSource;
  atlasUrl: string;
  rows: number;
  atlasWidth: number;
  atlasHeight: number;
  folder: string;
}

export interface ListedPet {
  id: string;
  name: string;
  source: PetSource;
  atlas_b64?: string | null;
  atlas_width?: number | null;
  atlas_height?: number | null;
  rows?: number | null;
  valid: boolean;
  warning?: string | null;
  folder: string;
}

export const bundledPet: PetCandidate = {
  id: "clawd",
  name: "Clawd",
  source: "bundled",
  atlasUrl: clawdAtlasUrl,
  rows: 9,
  atlasWidth: 1536,
  atlasHeight: 1872,
  folder: "clawd",
};

export function candidatesFromListedPets(pets: ListedPet[]): { candidates: PetCandidate[]; invalid: ListedPet[] } {
  const external = pets
    .filter((pet) => pet.source === "external" && pet.valid && pet.atlas_b64 && pet.rows && pet.atlas_width && pet.atlas_height)
    .map((pet) => ({
      id: pet.id,
      name: pet.name,
      source: "external" as const,
      atlasUrl: `data:image/webp;base64,${pet.atlas_b64}`,
      rows: pet.rows!,
      atlasWidth: pet.atlas_width!,
      atlasHeight: pet.atlas_height!,
      folder: pet.folder,
    }));
  return {
    candidates: [bundledPet, ...external],
    invalid: pets.filter((pet) => pet.source === "external" && !pet.valid),
  };
}

export function resolvePet(pets: readonly PetCandidate[], id: string | undefined): PetCandidate {
  return pets.find((pet) => pet.id === id) ?? bundledPet;
}

File: src-tauri/src/commands/pets.rs
use base64::Engine as _;
use serde::Serialize;
use std::fs;
use std::path::Path;

const MAX_ATLAS_BYTES: u64 = 5 * 1024 * 1024;

#[derive(Serialize)]
pub struct PetInfo {
    id: String,
    name: String,
    source: String,
    atlas_b64: Option<String>,
}

#[derive(serde::Deserialize)]
struct PetManifest {
    name: Option<String>,
}

fn external_pets(root: &Path) -> Vec<PetInfo> {
    let Ok(entries) = fs::read_dir(root) else {
        return Vec::new();
    };

    entries
        .flatten()
        .filter_map(|entry| {
            let file_type = entry.file_type().ok()?;
            if !file_type.is_dir() {
                return None;
            }
            let folder = entry.file_name().to_string_lossy().into_owned();
            let directory = entry.path();
            let atlas = ["spritesheet.webp", "spritesheet.png"]
                .into_iter()
                .map(|name| directory.join(name))
                .find(|path| path.is_file())?;
            let metadata = fs::metadata(&atlas).ok()?;
            if metadata.len() > MAX_ATLAS_BYTES {
                eprintln!("[pets] Skipping oversized atlas: {}", atlas.display());
                return None;
            }
            let bytes = fs::read(&atlas).ok()?;
            let name = fs::read_to_string(directory.join("pet.json"))
                .ok()
                .and_then(|contents| serde_json::from_str::<PetManifest>(&contents).ok())
                .and_then(|manifest| manifest.name)
                .filter(|name| !name.trim().is_empty())
                .unwrap_or_else(|| folder.clone());
            Some(PetInfo {
                id: format!("external:{folder}"),
                name,
                source: "external".to_string(),
                atlas_b64: Some(base64::engine::general_purpose::STANDARD.encode(bytes)),
            })
        })
        .collect()
}

#[tauri::command(async)]
pub async fn list_pets() -> Result<Vec<PetInfo>, String> {
    let mut pets = vec![PetInfo {
        id: "clawd".to_string(),
        name: "Clawd".to_string(),
        source: "bundled".to_string(),
        atlas_b64: None,
    }];
    let Some(home) = dirs::home_dir() else {
        return Ok(pets);
    };
    pets.extend(external_pets(&home.join(".codex").join("pets")));
    Ok(pets)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn external_pets_reads_manifest_name_and_png_atlas() {
        let root = tempfile::tempdir().unwrap();
        let pet_dir = root.path().join("sample");
        fs::create_dir(&pet_dir).unwrap();
        fs::write(pet_dir.join("spritesheet.png"), [1_u8, 2, 3]).unwrap();
        fs::write(pet_dir.join("pet.json"), r#"{"name":"Sample"}"#).unwrap();

        let pets = external_pets(root.path());
        assert_eq!(pets.len(), 1);
        assert_eq!(pets[0].id, "external:sample");
        assert_eq!(pets[0].name, "Sample");
        assert_eq!(pets[0].source, "external");
        assert_eq!(pets[0].atlas_b64.as_deref(), Some("AQID"));
    }
}
